crates/iota-indexer/src/backfill/ingestion/task.rs

```diff
@@ -0,0 +1,213 @@
+    },
+    db::ConnectionPool,
+    errors::IndexerError,
+};
+
+/// Orchestrates ingestion-driven backfill by buffering processed checkpoints
+/// and coordinating range-based commits.
+///
+/// `IngestionBackfillTask` spins up an ingestion workflow that continuously
+/// transforms checkpoints (via `Adapter<T>`), storing
+/// them in `ready_checkpoints`. Backfill operations can then drain these
+/// buffered records in order, pausing the backfill until the required
+/// checkpoint data arrives (via `notify`), and commit the chunks.
+pub struct IngestionBackfillTask<T: IngestionBackfillHandler> {
+    ready_checkpoints: Arc<DashMap<CheckpointSequenceNumber, Vec<T::ProcessedType>>>,
+    notify: Arc<Notify>,
+    _cancel_token: CancellationToken,
+}
+
+impl<T: IngestionBackfillHandler + 'static> IngestionBackfillTask<T> {
+    // Creates and starts a new ingestion‐driven backfill task using processor `T`.
+    #[expect(dead_code)]
+    pub(crate) async fn new(
+        remote_store_url: String,
+        start_checkpoint: CheckpointSequenceNumber,
+    ) -> Result<Self, IndexerError> {
+        let ready_checkpoints = Arc::new(DashMap::new());
+        let notify = Arc::new(Notify::new());
+        let adapter: Adapter<T> = Adapter {
+            ready_checkpoints: ready_checkpoints.clone(),
+            notify: notify.clone(),
+        };
+        let reader_options = ReaderOptions {
+            batch_size: 200,
+            ..Default::default()
+        };
+        let (executor, _cancel_token) = setup_single_workflow(
+            adapter,
+            remote_store_url,
+            start_checkpoint,
+            200,
+            Some(reader_options),
+        )
+        .await?;
+
+        tokio::spawn(async move {
+            if let Err(join_err) = executor.await {
+                error!(?join_err, "Ingestion executor panicked or was cancelled");
+            }
+        });
+
+        Ok(Self {
+            ready_checkpoints,
+            notify,
+            _cancel_token,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: IngestionBackfillHandler> BackfillTask for IngestionBackfillTask<T> {
+    async fn backfill_range(
+        &self,
+        pool: ConnectionPool,
+        range: &RangeInclusive<usize>,
+    ) -> Result<(), IndexerError> {
+        let mut processed_data = vec![];
+        let mut start = *range.start();
+        let end = *range.end();
+        loop {
+            while start <= end {
+                if let Some((_, processed)) = self
+                    .ready_checkpoints
+                    .remove(&(start as CheckpointSequenceNumber))
+                {
+                    processed_data.extend(processed);
+                    start += 1;
+                } else {
+                    break;
+                }
+            }
+            if start <= end {
+                self.notify.notified().await;
+            } else {
+                break;
+            }
+        }
```

Comment:
Seems to me that this could be simplified into:

```
        while start <= end {
            while let Some((_, processed)) = self
                .ready_checkpoints
                .remove(&(start as CheckpointSequenceNumber))
            {
                processed_data.extend(processed);
                start += 1;
            }

            if start <= end {
                self.notify.notified().await;
            }
        }
```

Or am I missing something?